src/ir/auth_logic/ast_construction.h

```diff
@@ -0,0 +1,60 @@
+
+/*
+ * Copyright 2022 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// This file contains the datalog IR (DLIR) which makes the translation from
+// this authorization logic into datalog simpler.
+
+#ifndef SRC_IR_AUTH_LOGIC_AST_CONSTRUCTION_H_
+#define SRC_IR_AUTH_LOGIC_AST_CONSTRUCTION_H_
+
+#include "src/ir/datalog/program.h"
+#include "src/ir/auth_logic/ast.h"
+#include "absl/status/statusor.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicParser.h"
+
+namespace raksha::ir::auth_logic {
+
+class Parser {
+public:
+
+//Parses an input string
+
+absl::StatusOr<Program> ParseProgram (absl::string_view prog_text);
+
+private:
+
+absl::StatusOr<Program> ConstructProgram (AuthLogicParser::ProgramContext* context);
+
+absl::StatusOr<Query> ConstructQuery (AuthLogicParser::QueryContext* context);
+
+absl::StatusOr<Fact> ConstructFact(AuthLogicParser::FactContext* context);
+
+absl::StatusOr<BaseFact> ConstructFlatFact(AuthLogicParser::FlatFactContext* context);
+
+absl::StatusOr<BaseFact> ConstructVerbphrase(Principal left_principal, AuthLogicParser::VerbphraseContext* context);
```

Comment:
If there is no conceivable way to recover from the errors that would be indicated by the `StatusOr` or if the errors should never come up, you can make these just return the wrapped types and instead use `CHECK` to fail if the checked state doesn't hold. In most of these cases, I think you're just returning `IllegalArgument` if your `dynamic_cast` doesn't work, so I think you can do this for all of these methods.